pubsub/redis.go

```diff
@@ -65,10 +75,79 @@ func (s *RedisSubscriber) Start() error {
 }
 
 func (s *RedisSubscriber) listen() error {
-	c, err := redis.DialURL(s.url)
 
-	if err != nil {
-		return err
+	var c redis.Conn
+	var err error
+
+	if s.sentinelsEnabled {
+
+		s.log.Debugf("Redis sentinel enabled")
+		s.log.Debugf("Redis sentinel parameters:  sentinels: %s,  masterName: %s", s.sentinels, s.masterName)
+		sentinels := strings.Split(s.sentinels, ",")
+		sntnl := &sentinel.Sentinel{
+			Addrs:      sentinels,
+			MasterName: s.masterName,
+			Dial: func(addr string) (redis.Conn, error) {
+				timeout := 500 * time.Millisecond
+
+				c, err := redis.Dial(
+					"tcp",
+					addr,
+					redis.DialConnectTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+				)
+				if err != nil {
+					s.log.Debugf("Failed to connect to sentinel %s", addr)
+					return nil, err
+				}
+				s.log.Debugf("Successfully connected to sentinel %s", addr)
+				return c, nil
+			},
+		}
+
+		defer sntnl.Close()
+
+		pool := &redis.Pool{
+			MaxIdle:     3,
+			MaxActive:   64,
+			Wait:        true,
+			IdleTimeout: 240 * time.Second,
+			Dial: func() (redis.Conn, error) {
+				masterAddr, err := sntnl.MasterAddr()
+				if err != nil {
+					s.log.Debugf("Failed to get master address from sentinel.")
+					return nil, err
+				}
+				s.log.Debugf("Got master address from sentinel: %s", masterAddr)
+				c, err := redis.Dial("tcp", masterAddr, redis.DialPassword(s.password))
+
+				if err != nil {
+					s.log.Debugf("Failed to connect to redis master %s", masterAddr)
+					return nil, err
+				}
+				s.log.Debugf("Successfully connected to redis master %s", masterAddr)
+				return c, nil
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				if !sentinel.TestRole(c, "master") {
+					return errors.New("Role check failed")
+				} else {
+					return nil
+				}
+			},
+		}
+
+		defer pool.Close()
+
+		c = pool.Get()
+
+	} else {
+		c, err = redis.DialURL(s.url)
```

Comment:
Can't we re-use this functionality by generating an URL from `sntnl.MasterAddr()`?